src/ast/passes/codegen_llvm.cpp

```diff
@@ -2799,10 +2799,38 @@ void CodegenLLVM::generateProbe(Probe &probe,
   if (probe.pred) {
     auto scoped_del = accept(probe.pred);
   }
+
   variables_.clear();
-  auto scoped_del = accept(probe.block);
+  if (current_attach_point_->expansion == ExpansionType::SESSION) {
+    BasicBlock *entry_probe = BasicBlock::Create(module_->getContext(),
+                                                 "entry_probe",
+                                                 func);
+    BasicBlock *exit_probe = BasicBlock::Create(module_->getContext(),
+                                                "exit_probe",
+                                                func);
+
+    Value *is_ret = CreateKernelFuncCall(Kfunc::bpf_session_is_return,
+                                         {},
+                                         "is_return");
+    b_.CreateCondBr(b_.CreateICmpNE(is_ret, b_.GetIntSameSize(0, is_ret)),
+                    exit_probe,
+                    entry_probe);
+
+    // Entry probe
+    b_.SetInsertPoint(entry_probe);
+    auto scoped_del_entry = accept(probe.block);
+    createRet();
 
-  createRet();
+    variables_.clear();
```

Comment:
Super nit: but might be good to have a codegen or runtime tests that exercises this `clear` as it's pretty important.

Reply:
Indeed, it looks like this is not even needed as (1) there's always just a single branch executed and (2) even if it was possible to execute both, since we have two separate probes, semantic analyser will make sure that each probe first assigns a scratch variable before reading it so the value from the other branch cannot get there.

I think that we would be fine using just one set of variables but it still feels a bit cleaner and safer to me to do the `clear()` here.

Reply:
I think if we have 2 reasons why it's not needed we should remove it (it's code smell) or add a comment as to why we're keeping it.

Reply:
If it ends up biting us in the ass, I'll buy you a beer 🍺